### ISSUE TITLE
Can we install docker on Windows CI machine?

### ISSUE THREAD

**Issue by yajiedesign** [apache/incubator-mxnet#18793]
Prepare to deploy CD in docker like Linux.
Can docker be installed on the machine and start Windows container mode?
leonard
@leezu 
@szha

**Comment by leezu** [comment]
Would this be a good opportunity to stop the "hot-patch the existing AMI (without having a mechanism to regenerate it)" and finalize the script to generate the AMI? What do you think @sandeep-krishnamurthy @ChaiBapchya @josephevans?

**Comment by marcoabreu** [comment]
Well it will still run Linux inside docker then, doesn't it? I haven't really been able to get much information about Windows containers

**Comment by leezu** [comment]
It should be possible to run windows inside the container, but cuda may not be supported @yajiedesign?

**Comment by yajiedesign** [comment]
Can run the Linux container.But there are some limitations.Can't run windows and Linux at the same time.Need to switch with command.
CUDA may be supported with docker desktop WSL 2 backend.

**Comment by yajiedesign** [comment]
At present, CUDA image production encountered difficulties. Some inexplicable errors, I am still trying

**Comment by yajiedesign** [comment]
@leezu  @szha I have successfully made cuda10 images, and completed the compilation and packaging in the docker container.
next we need to install docker in aws machine.

**Comment by marcoabreu** [comment]
Can you please show a demo before the team takes on the effort to update the ami?

**Comment by yajiedesign** [comment]
I try to gen a demo.

**Comment by yajiedesign** [comment]
I am push a domo image to dockerhub yajiedesign/mxnet_cd_demo:1.0
run c:\t\run_demo.bat start build.

**Comment by leezu** [comment]
Thank you @yajiedesign. Can you also share the Dockerfile for https://hub.docker.com/r/yajiedesign/mxnet_cd_demo

**Comment by yajiedesign** [comment]
https://github.com/yajiedesign/mxnet_cd_demo

**Comment by yajiedesign** [comment]
However, these dockerfiles cannot be used directly because they contain a large number of external files. The total is about 4GB.

**Comment by yajiedesign** [comment]
I am upload all file to onedriver https://1drv.ms/u/s!AgSFk5eCdAIXkVy8o_B54S1_n8ML?e=qkuJUu

**Comment by yajiedesign** [comment]
@szha

**Comment by leezu** [comment]
> these dockerfiles cannot be used directly because they contain a large number of external files.

Our CI actually needs to be able to build the container (at least that's how we do it for Linux containers). Should we request NVidia to provide containers a la https://hub.docker.com/r/nvidia/cuda/ for windows?

**Comment by josephevans** [comment]
I can spend some time creating Windows AMIs with docker installed. Can I use the existing AMIs we currently use for Windows CI? What version of docker and what installation steps are you using?

**Comment by yajiedesign** [comment]
@josephevans  I use ltsc2019 base image,You can choose the Windows version based on this page:https://docs.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/version-compatibility?tabs=windows-server-2004%2Cwindows-10-2004

the docker use the last version.I am use win10,so I donwload and run win10 docker setup and next next next etc.
win server look this :https://docs.microsoft.com/en-us/virtualization/windowscontainers/manage-docker/configure-docker-daemon

**Comment by yajiedesign** [comment]
@leezu if NVidia can provide windows cuda containers is the best.

**Comment by szha** [comment]
> @leezu if NVidia can provide windows cuda containers is the best.

cc @TristonC who tends to have magical ways of making such things happen.

**Comment by sandeep-krishnamurthy** [comment]
> I can spend some time creating Windows AMIs with docker installed. Can I use the existing AMIs we currently use for Windows CI? What version of docker and what installation steps are you using?

@josephevans - Can you please help on how to help @yajiedesign and get this work merged to CI?

**Comment by josephevans** [comment]
I don't think we want to re-use an existing Windows CI AMI for this, because it has a lot of stuff installed that may not be needed. Are you planning on installing all the build tools (MSVS, nvcc, etc) in the docker container instead of on the host system? Besides docker, what other software needs to be installed on the base AMI? Does NVidia drivers need to be installed?

**Comment by yajiedesign** [comment]
@josephevans For security reasons, I'm not going to run tests on CD machines. So maybe I just need to docker.jenkins slave,python.

**Comment by marcoabreu** [comment]
From an ami perspective, we are not differentiating between CD and CI machines. They all use the same image

**Comment by yajiedesign** [comment]
Then the NV driver and CUDA are needed to run the test, and the tool chain is still placed in the docker container.

**Comment by josephevans** [comment]
So just to recap, we need the following installed on a Windows AMI:

- docker
- python3
- jenkins slave run script (in python)
- java runtime environment (to run jenkins slave client)
- NVidia driver
- CUDA libraries/toolkit

Anything else?

**Comment by marcoabreu** [comment]
Just to make sure: you are aware of the scripts to automatically create a windows slave ami, right?

**Comment by josephevans** [comment]
@marcoabreu Yeah, I'm familiar with the scripts to create the AMIs. Thanks.

**Comment by szha** [comment]
what's the status for this? It's been almost three months.

**Comment by josephevans** [comment]
I have a new Windows AMI created with Cuda 11.0 and docker, but having issues with testing. Builds work, but unit tests are failing with: OSError: exception: access violation writing 0x0000000000000000

I believe this may be due to OpenCV or OpenBLAS dependencies.

**Comment by josephevans** [comment]
OK guys, I finally got new AMIs launched for Windows builds that include docker. They are currently in-place with Cuda 11.0 installed. You should be able to start migrating to containerized builds.
